pubsub/tests/unit/subscriber_test.py

```diff
@@ -0,0 +1,469 @@
+        )
+        for m in [mock1, mock2, mock3]:
+            await queue.put((m, 0.0))
+        await asyncio.sleep(0)
+        await asyncio.sleep(0)
+        await asyncio.sleep(0)
+        consumer_task.cancel()
+        mock1.assert_called_once()
+        mock2.assert_called_once()
+        mock3.assert_not_called()
+        assert queue.qsize() == 1
+        assert ack_queue.qsize() == 0
+
+    @pytest.mark.asyncio
+    async def test_consumer_drops_expired_messages(ack_deadline_cache,
+                                                   message,
+                                                   application_callback):
+        f = asyncio.Future()
+        f.set_result(0.0)
+        ack_deadline_cache.get = MagicMock(return_value=f)
+
+        queue = asyncio.Queue()
+        ack_queue = asyncio.Queue()
+        consumer_task = asyncio.ensure_future(
+            consumer(
+                queue,
+                application_callback,
+                ack_queue,
+                ack_deadline_cache,
+                1,
+                MagicMock()
+            )
+        )
+        await queue.put((message, 0.0))
+        await asyncio.sleep(0)
+        consumer_task.cancel()
+        application_callback.assert_not_called()
+        assert ack_queue.qsize() == 0
+        assert queue.qsize() == 0
+
+    @pytest.mark.asyncio
+    async def test_consumer_handles_callback_exception(
+        ack_deadline_cache, message
+    ):
+        queue = asyncio.Queue()
+        ack_queue = asyncio.Queue()
+        mock = MagicMock()
+
+        async def f(*args):
+            mock(*args)
+            raise RuntimeError
+
+        consumer_task = asyncio.ensure_future(
+            consumer(
+                queue,
+                f,
+                ack_queue,
+                ack_deadline_cache,
+                1,
+                MagicMock()
+            )
+        )
+        await queue.put((message, 0.0))
+        await asyncio.sleep(0)
+        await asyncio.sleep(0)
+        consumer_task.cancel()
+        mock.assert_called_once()
+        assert ack_queue.qsize() == 0
+        assert queue.qsize() == 0
+
+    # ========
+    # acker
+    # ========
+
+    @pytest.mark.asyncio
+    async def test_acker_does_ack(subscriber_client):
+        queue = asyncio.Queue()
+        acker_task = asyncio.ensure_future(
+            acker(
+                'fake_subscription',
+                queue,
+                subscriber_client,
+                0.0,
+                MagicMock()
+            )
+        )
+        await queue.put('ack_id')
+        await asyncio.sleep(0)
+        acker_task.cancel()
+        subscriber_client.acknowledge.assert_called_once_with(
+            'fake_subscription', ack_ids=['ack_id'])
+        assert queue.qsize() == 0
+
+    @pytest.mark.asyncio
+    async def test_acker_handles_exception(subscriber_client):
+        async def f(*_args, **_kwargs):
+            await asyncio.sleep(0)
+            raise RuntimeError
```

Comment:
Similar to my above comment - I don't think we get cancelled before raising `RuntimeError`

Reply:
Looks good now!